keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+TBD
+
+## Motivation
+
+TBD
+
+### Goals
+
+TBD
+
+### Non-Goals
+
+TBD
+
+### Future Goals
+
+TBD
+
+## Proposal
+
+TBD
+
+### Implementation Details/Notes/Constraints
+
+more details on the original doc at [https://docs.google.com/document/d/1xw5m71SKcP4wzeIigWv3wZ3_N9w9G0xHYBJUpxF034w/edit](https://docs.google.com/document/d/1xw5m71SKcP4wzeIigWv3wZ3_N9w9G0xHYBJUpxF034w/edit)
+
+##### Definitions
+
+- *BucketBinding*: a set of rules defining how we bind an inbound request w/ a bucket.
+It works similar to \[Cluster\]RoleBinding in the RBAC world.
+- *Quota*: a request must acquire one quota before its execution. Basically,
+we have two kinds of quota: reserved and shared. With reserved quota, the
+request can be executed immediately. If the request fails to acquire reserved
+quota from the bucket, it will be queued until a shared quota is distributed
+before its execution.
+- *Bucket*: a quota pool. The quota will be taken from the bucket when it's
+acquired by any request and will be returned to the bucket when the request
+finishes. A bucket has following attributes:
+  - *Reserved Quota*: an initial guaranteed concurrency can be provided by
+  the bucket
+  - *Shared Quota*: an initial shared concurrency can be provided by the bucket.
+  Shared quota cannot be consumed by requests queued at lower priorities
+  than the bucket's.
+  - *Weight*: only works for shared quota. The higher weight a bucket has,
+  the more chance the requests hitting the buckets can be executed from the
+  queue.
+  - *Priority*: a fixed/preset set of priorities ranged 1...N. The lower
+  number means higher priority. Each priority has one or more buckets providing
+  quotas and buckets from higher priority can "borrow" shared quotas from
+  the lowers.
+- *Extra Shared Quota / ESQ*: global shared quota provided by a logical
```

Comment:
just reading top to bottom, it's not obvious why this isn't simply a normal bucket with reserved=0 and "all the rest" shared quota.